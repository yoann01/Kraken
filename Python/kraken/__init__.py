"""Kraken Framework."""

import logging
import os

__all__ = ['core', 'helpers', 'plugins', 'ui']


krakenPath = os.path.abspath(os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', '..'))
if os.environ.get('KRAKEN_PATH', None) is None:
    os.environ['KRAKEN_PATH'] = krakenPath

krakenExtsPath = os.path.join(krakenPath, 'Exts')
if 'FABRIC_EXTS_PATH' in os.environ:
    if krakenExtsPath not in os.environ['FABRIC_EXTS_PATH']:
        os.environ['FABRIC_EXTS_PATH'] = krakenExtsPath + os.pathsep + os.environ['FABRIC_EXTS_PATH']

krakenPresetsPath = os.path.join(krakenPath, 'Presets', 'DFG')
if 'FABRIC_DFG_PATH' in os.environ:
    if krakenPresetsPath not in os.environ['FABRIC_DFG_PATH']:
        os.environ['FABRIC_DFG_PATH'] = krakenPresetsPath + os.pathsep + os.environ['FABRIC_DFG_PATH']


logging.basicConfig(format='[KRAKEN] %(levelname)s: %(message)s', level=logging.INFO)

# Custom inform level for use with UI label getting added to the status bar.
logging.INFORM = 25
logging.addLevelName(logging.INFORM, 'INFORM')
logging.Logger.inform = lambda inst, msg, *args, **kwargs: inst.log(logging.INFORM, msg, *args, **kwargs)


def release():
    print "RELEASE THE KRAKEN!!!!"
