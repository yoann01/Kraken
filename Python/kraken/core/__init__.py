"""Kraken Core."""

VERSION_MAJOR = 1
VERSION_MINOR = 2
VERSION_BUILD = 6
VERSION_SUFFIX = ""


def getVersion():
    """Contatenates the version globals and returns the current version of
    Kraken.

    Returns:
        str: Current version of Kraken.

    """

    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString
