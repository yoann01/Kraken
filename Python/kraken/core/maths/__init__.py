"""Kraken - math module."""

from math_object import MathObject
from vec2 import Vec2
from vec3 import Vec3
from vec4 import Vec4
from quat import Quat
from euler import Euler
from xfo import Xfo
from mat33 import Mat33
from mat44 import Mat44
from rotation_order import RotationOrder
from color import Color
from constants import DEG_TO_RAD, RAD_TO_DEG, PI


def Math_radToDeg(val):
    """Converts radians to degrees.

    Args:
        val (float): value to convert to degrees.

    Returns:
        float: Value in degrees.

    """

    return val * RAD_TO_DEG

def Math_degToRad(val):
    """Converts degrees to radians.

    Args:
        val (float): value to convert to radians.

    Returns:
        float: Value in radians.

    """

    return val * DEG_TO_RAD


def decodeValue(jsonData):
    """Returns a constructed math value based on the provided json data.

    Args:
        jsondata (dict): The JSON data to use to decode into a Math value.

    Returns:
        object: The constructed math value

    """

    if type(jsonData) is not dict:
        return jsonData

    if '__mathObjectClass__' not in jsonData:
        raise Exception("Invalid JSON data for constructing value:" + str(jsonData))

    if jsonData['__mathObjectClass__'] == 'Vec2':
        val = Vec2()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Vec3':
        val = Vec3()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Vec4':
        val = Vec4()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Euler':
        val = Euler()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Quat':
        val = Quat()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Xfo':
        val = Xfo()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Mat33':
        val = Mat33()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Mat44':
        val = Mat44()
        val.jsonDecode(jsonData, decodeValue)
    else:
        raise Exception("Unsupported Math type:" + jsonData['__mathObjectClass__'])

    return val
