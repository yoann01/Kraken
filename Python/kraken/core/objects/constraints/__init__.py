"""Kraken Constraints."""

__all__ = ['constraint',
           'scale_constraint',
           'orientation_constraint',
           'position_constraint',
           'pose_constraint']