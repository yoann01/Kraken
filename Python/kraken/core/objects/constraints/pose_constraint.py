"""Kraken - objects.Constraints.PoseConstraint module.

Classes:
PoseConstraint - Pose Constraint.

"""

from constraint import Constraint
from kraken.core.maths.xfo import Xfo
from kraken.core.maths.vec3 import Vec3
from kraken.core.maths.quat import Quat


class PoseConstraint(Constraint):
    """Pose Constraint."""

    def __init__(self, name, metaData=None):
        super(PoseConstraint, self).__init__(name, metaData=metaData)

