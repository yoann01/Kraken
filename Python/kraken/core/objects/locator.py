"""Kraken - objects.locator module.

Classes:
Locator -- Locator representation.

"""

from kraken.core.objects.object_3d import Object3D


class Locator(Object3D):
    """Locator object."""

    def __init__(self, name, parent=None, flags=None, metaData=None):
        super(Locator, self).__init__(name, parent=parent, flags=flags, metaData=metaData)
