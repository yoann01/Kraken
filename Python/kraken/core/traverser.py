"""Kraken - traverser module.

Classes:
Traverser - Base Traverser.

"""

from kraken.core.objects.scene_item import SceneItem
from kraken.core.objects.object_3d import Object3D
from kraken.core.objects.components.component import Component
from kraken.core.objects.attributes.attribute_group import AttributeGroup
from kraken.core.objects.constraints.constraint import Constraint
from kraken.core.objects.operators.operator import Operator


class Traverser(object):
    """Kraken base traverser for any scene item.

    The traverser iterates through all root items and determines the
    dependencies between objects, building an ordered list as it goes. This
    order is then used by the builder to ensure that objects are created and
    evaluated in the correct order. Offset's will then be reliable.

    """

    def __init__(self, name='Traverser'):
        self._rootItems = []
        self.reset()

    # ==================
    # Property Methods
    # ==================
    @property
    def rootItems(self):
        """Gets the root items of this Traverser.

        Returns:
            list: The root items of this Traverser.

        """

        return self._xfo

    def addRootItem(self, item):
        """Adds a new root object to this Traverser

        Args:
            item (SceneItem): The SceneItem to add as a root item.

        Returns:
            bool: True if successful.

        """

        for rootItem in self._rootItems:
            if rootItem.getId() == item.getId():
                return False

        self._rootItems.append(item)

        return True

    def addRootItems(self, items):
        """Adds a bunch of root items to this Traverser

        Args:
            items (SceneItem[]): The SceneItems to add as root items.

        Returns:
            bool: True if successful.

        """

        for item in items:
            self.addRootItem(item)

        return True

    @property
    def items(self):
        """Gets the traversed items of this Traverser.

        Returns:
            list: The traversed items of this Traverser.

        """

        return self._items

    def getItemsOfType(self, typeNames):
        """Gets only the traversed items of a given type.

        Args:
            typeName (str): The name of the type to look for.

        Returns:
            list: The items in the total items list matching the given type.

        """

        if not isinstance(typeNames, list):
            typeNames = [typeNames]

        result = []
        for item in self._items:
            if item.isOfAnyType(typeNames):
                result.append(item)

        return result

    # =============
    # Traverse Methods
    # =============
    def reset(self):
        """Resets all internal structures of this Traverser."""

        self._visited = {}
        self._items = []

    def traverse(self, itemCallback=None, discoverCallback=None,
                 discoveredItemsFirst=True):
        """Visits all objects within this Traverser based on the root items.

        Args:
            itemCallback (func): A callback function to be invoked for each
                item visited.
            discoverCallback (func): A callback to return an array of children
                for each item.

        """

        self.reset()

        if discoverCallback is None:
            discoverCallback = self.discoverBySource

        for item in self._rootItems:
            self.__visitItem(item,
                             itemCallback,
                             discoverCallback,
                             discoveredItemsFirst)

        return self.items

    def __collectVisitedItem(self, item, itemCallback):
        """Doc String.

        Args:
            item (Type): information.
            itemCallback (Type): information.

        """

        if itemCallback is not None:
            itemCallback(item=item, traverser=self)

        self._items.append(item)

    def __visitItem(self, item, itemCallback, discoverCallback, discoveredItemsFirst):
        """Doc String.

        Args:
            item (Type): information.
            itemCallback (Type): information.
            discoverCallback (Type): information.
            discoveredItemsFirst (Type): information.

        Returns:
            Type: information.

        """

        if self._visited.get(item.getId(), False):
            return False

        self._visited[item.getId()] = True

        if hasattr(item, 'getParent') and item.getParent():
            # If this is an attribute and we have not traversed its parent AttributeGroup then skip this
            # and visit the parent so we get this attribute and all others from there (for the sake of attr order)
            if item.isTypeOf("Attribute") and not self._visited.get(item.getParent().getId(), False):
                self._visited[item.getId()] = False
                self.__visitItem(item.getParent(),
                             itemCallback,
                             discoverCallback,
                             discoveredItemsFirst)
                return False

            self.__visitItem(item.getParent(),
                             itemCallback,
                             discoverCallback,
                             discoveredItemsFirst)

        sourcedByConstraintOrOperator = False
        if discoveredItemsFirst:
            for source in item.getSources():
                if isinstance(source, (Constraint, Operator)):
                    sourcedByConstraintOrOperator = True
                    break

        if not discoveredItemsFirst or sourcedByConstraintOrOperator:
            self.__collectVisitedItem(item, itemCallback)

        if discoverCallback:
            if isinstance(item, AttributeGroup):
                discoveredItems = self.discoverChildren(item)
            else:
                discoveredItems = discoverCallback(item)

            if discoveredItems:
                for discoveredItem in discoveredItems:
                    self.__visitItem(discoveredItem,
                                     itemCallback,
                                     discoverCallback,
                                     discoveredItemsFirst)

        if discoveredItemsFirst and not sourcedByConstraintOrOperator:
            self.__collectVisitedItem(item, itemCallback)

        return True

    def discoverChildren(self, item):
        """Doc String.

        Args:
            item (Type): information.

        Returns:
            Type: information.

        """

        result = []

        if isinstance(item, Component):
            subItems = item.getItems().values()
            for subItem in subItems:
                if isinstance(subItem, AttributeGroup):
                    continue
                result.append(subItem)

        elif isinstance(item, Object3D):
            for i in xrange(item.getNumAttributeGroups()):
                result.append(item.getAttributeGroupByIndex(i))
            result += item.getChildren()

        if isinstance(item, AttributeGroup):
            for i in xrange(item.getNumAttributes()):
                result.append(item.getAttributeByIndex(i))

        return result

    def discoverBySource(self, item):
        """Doc String.

        Args:
            item (Type): information.

        Returns:
            Type: information.

        """

        result = []

        for source in item.getSources():

            if isinstance(source, Operator):
                for outputName in source.getOutputNames():
                    operatorOutputs = source.getOutput(outputName)
                    if not isinstance(operatorOutputs, list):
                        operatorOutputs = [operatorOutputs]
                    for operatorOutput in operatorOutputs:
                        if not isinstance(operatorOutput, SceneItem):
                            continue
                        result.append(operatorOutput)

            result.append(source)

        return result
