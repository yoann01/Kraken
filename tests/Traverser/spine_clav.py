from kraken import plugins
from kraken.core.maths import Vec3
from kraken.core.objects.object_3d import Object3D
from kraken.core.objects.locator import Locator
from kraken.core.objects.constraints.pose_constraint import PoseConstraint
from kraken.core.traverser.traverser import Traverser
from kraken_examples.spine_component import SpineComponentRig
from kraken_examples.clavicle_component import ClavicleComponentRig

spineComponent = SpineComponentRig('spine')
spineComponent.loadData(data={
    'cogPosition': Vec3(0.0, 11.1351, -0.1382),
    'spine01Position': Vec3(0.0, 11.1351, -0.1382),
    'spine02Position': Vec3(0.0, 11.8013, -0.1995),
    'spine03Position': Vec3(0.0, 12.4496, -0.3649),
    'spine04Position': Vec3(0.0, 13.1051, -0.4821),
    'numDeformers': 4
})
clavComponent = ClavicleComponentRig('clavicle')

vertebraeOutputs = spineComponent.getOutputByName('spineVertebrae')
clavicleSpineEndInput = clavComponent.getInputByName('spineEnd')
clavicleSpineEndInput.setConnection(vertebraeOutputs, 2)

trav = Traverser()
trav.addRootItem(spineComponent)
trav.addRootItem(clavComponent)

items = trav.traverse(discoverCallback = trav.discoverChildren)

trav.reset()
trav.addRootItems(items)

def callback(**args):
    item = args.get('item', None)
    print 'Visited '+item.getDecoratedPath()

trav.traverse(itemCallback = callback)
