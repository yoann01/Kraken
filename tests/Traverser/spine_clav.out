Failed to find DCC client. Falling back to Python client.
Visited spine:M
Visited clavicle:M
Visited controls
Visited controls.spine:M
Visited controls.spine:M.inputs
Visited controls.spine:M.inputs.inputs
Visited controls.spine:M.inputs.inputs.drawDebug
Visited controls.spine:M.inputs.inputs.rigScale
Visited controls.spine:M.inputs.inputs.length
Visited controls.spine:M.inputs.mainSrt
Visited controls.spine:M.outputs
Visited controls.spine:M.outputs.outputs
Visited controls.spine:M.outputs.cog
Visited controls.spine:M.cog
Visited controls.spine:M.cog.cog_To_mainSrt
Visited controls.spine:M.cog.cog
Visited controls.spine:M.outputs.cog.cog_To_cog
Visited controls.spine:M.outputs.spineBase
Visited controls.spine:M.outputs.spine01
Visited controls.spine:M.outputs.spine02
Visited controls.spine:M.outputs.spine03
Visited controls.spine:M.outputs.spine04
Visited controls.spine:M.cog.cog.spine01
Visited controls.spine:M.cog.cog.spine01.spine01
Visited controls.spine:M.cog.cog.spine01.spine01.spine02
Visited controls.spine:M.cog.cog.spine01.spine01.spine02.spine02
Visited controls.spine:M.cog.cog.spine01.spine01.spine02.spine02.spine03
Visited controls.spine:M.cog.cog.spine01.spine01.spine02.spine02.spine03.spine03
Visited controls.spine:M.cog.cog.spine04
Visited controls.spine:M.cog.cog.spine04.spine04
Visited spine:M.spineKLOp
Visited controls.spine:M.outputs.spineBase.spineBase_To_spineBase
Visited controls.spine:M.outputs.pelvis
Visited controls.spine:M.cog.cog.pelvis
Visited controls.spine:M.cog.cog.pelvis.pelvis
Visited controls.spine:M.outputs.pelvis.pelvis_To_pelvis
Visited controls.spine:M.outputs.spineEnd
Visited controls.spine:M.outputs.spineEnd.spineEnd_To_spineEnd
Visited deformers.spine:M.spine01
Visited deformers
Visited deformers.spine:M
Visited deformers.spine:M.spine02
Visited deformers.spine:M.spine03
Visited deformers.spine:M.spine04
Visited spine:M.spineDeformerKLOp
Visited deformers.spine:M.pelvis
Visited spine:M.pelvisDeformerKLOp
Visited controls
Visited controls.clavicle:M
Visited controls.clavicle:M.inputs
Visited controls.clavicle:M.inputs.inputs
Visited controls.clavicle:M.inputs.inputs.drawDebug
Visited controls.clavicle:M.inputs.inputs.rigScale
Visited controls.clavicle:M.inputs.inputs.rightSide
Visited controls.clavicle:M.inputs.spineEnd
Visited controls.clavicle:M.inputs.spineEnd.spineEnd_To_spine03_PoseConstraint
Visited controls.clavicle:M.outputs
Visited controls.clavicle:M.outputs.outputs
Visited controls.clavicle:M.outputs.clavicle
Visited controls.clavicle:M.clavicle
Visited controls.clavicle:M.clavicle.clavicle_To_spineEnd
Visited controls.clavicle:M.clavicle.clavicle
Visited controls.clavicle:M.outputs.clavicle.clavicle_To_clavicle
Visited controls.clavicle:M.outputs.clavicleEnd
Visited controls.clavicle:M.outputs.clavicleEnd.clavicleEnd_To_clavicle
Visited deformers.clavicle:M.clavicle
Visited deformers
Visited deformers.clavicle:M
Visited clavicle:M.clavicleDeformerKLOp
